File: 1043/B/prog.cpp
#include <bits/stdc++.h>


template <class Class>
void Print(Class instance) {
    std::cout << instance << std::endl;
}

void Print() {
    std::cout << std::endl;
}

std::vector<size_t> ComputeZ(const std::vector<int64_t>& string) {
    std::vector<size_t> result(string.size());
    size_t left = 0;
    size_t right = 0;
    for (size_t index = 0; index < string.size(); ++index) {
        if (right > index) {
            result[index] = std::min(right - index, result[index - left]);
        } else {
            result[index] = 0;
        }
        while (index + result[index] < string.size()
                && string[index + result[index]] == string[result[index]])
        {
            ++result[index];
        }
        if (right < index + result[index]) {
            right = index + result[index];
            left = index;
        }
    }
    return result;
}


int main() {
    std::ios_base::sync_with_stdio(false);
    #ifdef LOCAL
        std::ifstream input("input.txt");
        std::ofstream output("output.txt");
    #else
        auto& input = std::cin;
        auto& output = std::cout;
    #endif
    size_t n;
    input >> n;
    std::vector<int64_t> x(n);
    int64_t prev = 0;
    for (size_t index = 0; index < n; ++index) {
        input >> x[index];
        int64_t prev_prev = prev;
        prev = x[index];
        x[index] -= prev_prev;
    }
    std::vector<size_t> z = ComputeZ(x);
    std::vector<size_t> answer;
    for (size_t index = 1; index < n; ++index) {
        if (z[index] + index == n) {
            answer.push_back(index);
        }
    }
    answer.push_back(n);

    output << answer.size() << std::endl;
    for (size_t index : answer) {
        output << index << ' ';
    }

    return 0;
}

File: 1043/B/prog.py
import sys


def compute_z(x):
    z = [0 for i in x]
    left = 0
    right = 0
    for i in range(len(x)):
        z[i] = max(0, min(right - i, z[i - left]))
        while i + z[i] < len(x) and x[z[i]] == x[i + z[i]]:
            z[i] += 1
        if i + z[i] > right:
            right = i + z[i]
            left = i
    return z


def main(inp, out):
    n = int(inp.readline())
    a = list(map(int, inp.readline().split()))
    prev = 0
    x = [0 for i in range(n)]
    for i in range(n):
        x[i] = a[i] - prev
        prev = a[i]
    z = compute_z(x)
    answer = []
    for i in range(1, n):
        if z[i] + i == n:
            answer.append(i)
    answer.append(n)
    print(len(answer), file=out)
    print(' '.join(str(el) for el in answer), file=out)


if __name__ == '__main__':
    if sys.platform == 'win32':
        main(sys.stdin, sys.stdout)
    else:
        with open("input.txt", 'r') as inp, open("output.txt", 'w') as out:
            main(inp, out)
